src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala

```diff
@@ -88,43 +113,80 @@ class LeonardoService(protected val dataprocConfig: DataprocConfig,
-          case Some(key) => googleIamDAO.removeServiceAccountKey(WorkbenchGoogleProject(dataprocConfig.leoGoogleProject.string), cluster.googleServiceAccount, key)
-          case None => Future.successful(())
+  def deleteCluster(userInfo: UserInfo, googleProject: GoogleProject, clusterName: ClusterName): Future[Int] = {
+    for {
+      cluster <- getActiveClusterDetails(userInfo, googleProject, clusterName) //throws 404 if no permissions
+      count <- internalDeleteCluster(cluster)
+    } yield { count }
+  }
+
+  def internalDeleteCluster(cluster: Cluster): Future[Int] = {
+    if(cluster.status.isActive) {
+      // Delete the service account key in Google, if present
+      val deleteServiceAccountKey = dbRef.inTransaction { dataAccess =>
+        dataAccess.clusterQuery.getServiceAccountKeyId(cluster.googleProject, cluster.clusterName)
+      } flatMap {
+        case Some(key) => googleIamDAO.removeServiceAccountKey(WorkbenchGoogleProject(dataprocConfig.leoGoogleProject.string), cluster.googleServiceAccount, key)
+        case None => Future.successful(())
+      }
+
+      for {
+        _ <- deleteServiceAccountKey
+        _ <- gdDAO.deleteCluster(cluster.googleProject, cluster.clusterName)
+        recordCount <- dbRef.inTransaction(dataAccess => dataAccess.clusterQuery.markPendingDeletion(cluster.googleId))
+      } yield {
+        clusterMonitorSupervisor ! ClusterDeleted(cluster)
+        recordCount
+      }
+    } else Future.successful(0)
+  }
+
+  def listClusters(userInfo: UserInfo, params: LabelMap): Future[Seq[Cluster]] = {
+    for {
+      paramMap <- processListClustersParameters(params)
+      clusterList <- dbRef.inTransaction { da => da.clusterQuery.listByLabels(paramMap._1, paramMap._2) }
+
+      //Filter by authorization.
+      //Set of projects these clusters are in
+      clusterProjects = clusterList.map { _.googleProject }.toSet
+      //Does the user have god-mode, i.e. can they see all clusters in all projects?
+      hasAllProjectsPermission <- authProvider.hasPermissionInAllProjects(userInfo.userEmail, ListClusters)
+      //Get the list of projects with list-clusters permission
+      projectsWithListClusters <- if( hasAllProjectsPermission ) {
+          Future.successful(clusterProjects)
+        } else {
+          authProvider.getProjectsWithListClustersPermission(userInfo.userEmail)
         }
 
-        for {
-          _ <- deleteServiceAccountKey
-          _ <- gdDAO.deleteCluster(googleProject, clusterName)
-          recordCount <- dbRef.inTransaction(dataAccess => dataAccess.clusterQuery.markPendingDeletion(cluster.googleId))
-        } yield {
-          clusterMonitorSupervisor ! ClusterDeleted(cluster)
-          recordCount
+      //Find out which clusters we'll need to check one-by-one and which are visible by dint of project-level or god permissions.
+      (visibleClusters, checkHarder) = clusterList.partition( c => projectsWithListClusters.contains(c.googleProject) )
```

Comment:
I would probably prefer to do the filtering in the database query, rather than Scala code.

Reply:
How? It's possible to have permissions for a cluster but not its owning project. I think that means we have to get them all and check them one by one.

Reply:
Ah, I should actually read the code. You're getting all clusters, and then taking the union of "visible with project-level permissions" and "visible with notebook-level permissions". Right?

One other idea: instead of having `LeoAuthProvider.getProjectsWithListClustersPermission`, could we just filter the list of clusters on `(LeoAuthProvider.hasNotebookClusterPermission(email, GetClusterDetails, cluster.googleId) || LeoAuthProvider. hasProjectPermission(email, ListClusters, cluster.googleProject)`? Not real code but you see what I mean?

Reply:
Yes, but now imagine I'm a project owner and my comp bios are busy bees making lots of dataprocs. Two calls to Sam per cluster... ultimately this is a performance optimization (premature I know!).

Reply:
That said, perhaps clean interface >> "you do your own caching if your thing is slow".

Reply:
I would vote for cleaner APIs and optimize later. :) We could optimize the Sam calls to be O(num projects + num clusters). Also can the Sam impl easily support the “getProjects...” api? It might involve an LDAP search which could be slow itself.

Reply:
Doug has promised me an "get all resources I have X action on" endpoint; I hadn't thought much about it beyond that. I am convinced -- weird API will go.